### PR TITLE
[improvements] arrows x enclosing shapes x precision.

### PR DESCRIPTION
**Pull request by steveruizok**
This PR makes several improvements to the behavior of arrows as they relate to precision and container relationships.

- an arrow's terminals are always "true" and are never snapped to { x: .5, y: .5 } as they were previously when not precise
- instead, a new `isPrecise` boolean is added to the arrow terminal
- when an arrow terminal renders "imprecisely" it will be placed to the center of the bound shape
- when an arrow terminal renders "precisely" it will be placed at the normalized location within the bound shape

Kapture 2023-11-29 at 23 12 12

The logic now is...
- if the user has indicated precision by "pausing" while drawing the arrow, it will be precise
- otherwise...
- if both of an arrow's terminals are bound to the same shape, both will be precise
- if a terminal is bound to a shape that contains the shape that its opposite terminal is bound to, it will be precise
- if a terminal is bound to a shape that contains the shape that its opposite terminal is bound to, it will be precise
- or else it will be imprecise

If the spatial relationships change, the precision may change as well.

Fixes https://github.com/tldraw/tldraw/issues/2204

Note: a previous version of this PR was based around ancestry but that's not actually important.

### Change Type

- [x] `minor` — New feature

### Test Plan

1. Draw an arrow between a frame and its descendant
2. Draw an arrow inside of a shape to another shape contained within the bounds of the big shape
3. Vis versa
4. Vis versa

- [x] Unit Tests

### Release Notes

- Improves the logic about when to draw "precise" arrows between the center of bound shapes.

### CODE REVIEW

**Comment by steveruizok** [comment]
Kapture 2023-11-29 at 23 28 13

Still bugs with enclosing shapes

Kapture 2023-11-29 at 23 25 05

File: packages/editor/src/lib/editor/Editor.ts [hunk 1]
```diff
@@ -4353,9 +4353,10 @@ export class Editor extends EventEmitter<TLEventMap> {
 	/** @internal */
 	@computed private _getShapeMaskCache(): ComputedCache<Vec2d[], TLShape> {
 		return this.store.createComputedCache('pageMaskCache', (shape) => {
-			if (isPageId(shape.parentId)) {
-				return undefined
-			}
+			if (isPageId(shape.parentId)) return undefined
+
+			const util = this.getShapeUtil(shape)
+			if (!util.canBeMasked(shape)) return undefined
```
[the remark this answers is not in the document]
Comment (ds300) [reply packages/editor/src/lib/editor/Editor.ts:4359]: Kapture 2023-12-01 at 12 08 16

File: packages/tldraw/src/lib/shapes/arrow/ArrowShapeUtil.tsx [hunk 2]
```diff
@@ -302,21 +292,36 @@ export class ArrowShapeUtil extends ShapeUtil<TLArrowShape> {
...
+			// Turn off precision if we're within a certain distance to the center of the shape.
+			// Funky math but we want the snap distance to be 4 at the minimum and either
+			// 16 or 15% of the smaller dimension of the target shape, whichever is smaller
+			if (
+				Vec2d.Dist(pointInTargetSpace, targetBounds.center) <
+				Math.max(4, Math.min(Math.min(targetBounds.width, targetBounds.height) * 0.15, 16)) /
+					this.editor.getZoomLevel()
+			) {
+				normalizedAnchor.x = 0.5
+				normalizedAnchor.y = 0.5
```
[the remark this answers is not in the document]
Comment (steveruizok) [reply packages/tldraw/src/lib/shapes/arrow/ArrowShapeUtil.tsx:316]: No, in this case we would actually want to snap the center even if precise was true. :psyduck:

File: packages/editor/src/lib/editor/Editor.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (steveruizok) [reply packages/editor/src/lib/editor/Editor.ts:4359]: That's a good catch. I'm going to pull this part of the PR, though I think it improves the UX in the more common circumstances.

**Pull request merged**